File: spreadsheet_conversion/blood/pmod/convert_pmod_to_blood.py
import pandas as pd
import warnings
import re
from pathlib import Path


class PmodToBlood:
    def __init__(
            self,
            whole_blood_activity: Path,
            parent_fraction: Path,
            plasma_activity: Path = None,
            blood_sampling_type: str = 'both',
            output_name: str = '',
            output_json: bool = False,
            **kwargs):

        # whole blood and parent fraction are required, always attempt to load
        self.blood_series = {'whole_blood_activity': self.load_pmod_file(whole_blood_activity),
                             'parent_fraction': self.load_pmod_file(parent_fraction)}

        # plasma activity is not required, but is used if provided
        if plasma_activity:
            self.blood_series['plasma_activity'] = self.load_pmod_file(plasma_activity)
        self.data_collection = {}


        # one may encounter data collected manually and/or automatically, we vary our logic depending on the case
        self.data_collection = {}

        for blood_sample in self.blood_series.keys():
            var = f"{blood_sample}_collection_method"
            if not kwargs.get(var, None):
                self.ask_recording_type(blood_sample)
            else:
                self.data_collection[blood_sample] = kwargs.get(var)
        
        # scale time to seconds rename columns
        self.scale_time_rename_columns()

        # check blood files for consistency
        self.check_time_info()

        if output_name:
            self.output_name = Path(output_name)
        else:
            self.output_name = Path(whole_blood_activity).parent

    @staticmethod
    def load_pmod_file(pmod_blood_file: Path):
        if pmod_blood_file.is_file() and pmod_blood_file.exists():
            loaded_file = pd.read_excel(str(pmod_blood_file))
            return loaded_file
        else:
            raise FileNotFoundError(str(pmod_blood_file))

    def check_time_info(self):
        """
        Checks for time units, and time information between .bld files, number of rows and the values
        in the time index must be the same across each input .bld file. Additionally, renames time column
        to 'time' instead of what it's defined as in the pmod file.
        """
        # if there is only a single input do nothing, else go through each file. This shouldn't get reached
        # as whole_blood_activity and plasma_activity are required
        if len(self.blood_series) >= 2 and len(set(self.data_collection.values())) ==1:
            row_lengths = {}
            for key, bld_data in self.blood_series.items():
                row_lengths[key] = len(bld_data)

            if len(set(row_lengths.values())) > 1:
                err_message = f"Sampling method for all PMOD blood files (.bld) given as " \
                              f"{list(set(self.data_collection.values()))[0]} must be of the same dimensions row-wise!\n"
                for key, value in row_lengths.items():
                    err_message += f"{key} file has {value} rows\n"

                err_message += "Check input files are valid."

                raise Exception(err_message)

            # lastly make sure the same time points exist across each input file/dataframe
            whole_blood_activity = self.blood_series.pop('whole_blood_activity')
            for key, dataframe in self.blood_series.items():
                try:
                    assert whole_blood_activity['time'].equals(dataframe['time'])
                except AssertionError:
                    raise AssertionError(f"Time(s) must have same values between input files, check time columns.")
            # if it all checks out put the whole blood activity back into our blood series object
            self.blood_series['whole_blood_activity'] = whole_blood_activity
        
        # checks to make sure that an autosampled file has more entries in it than a manually sampled file, John Henry must lose.
        elif len(self.blood_series) >= 2 and len(set(self.data_collection.values())) > 1:
            # check to make sure auto sampled .bld files have more entries than none autosampled
            compare_lengths = []
            for key, sampling_type in self.data_collection.items():
                compare_lengths.append({'name': key,'sampling_type': sampling_type, 'sample_length': len(self.blood_series[key])})

            auto_sampled = []
            manually_sampled = []
            for each in compare_lengths:
                if 'auto' in str.lower(each['sampling_type']):
                    auto_sampled.append(each)
                elif 'manual' in str.lower(each['sampling_type']):
                    manually_sampled.append(each)
            print(f"auto_sampled: {auto_sampled}")
            print(f"manually_sampled: {manually_sampled}")

            for auto in auto_sampled:
                for manual in manually_sampled:
                    if auto['sample_length'] < manual['sample_length']:
                        warnings.warn(
                            f"Autosampled .bld input for {list(auto.keys())[0]} has {len(auto['sample_length'])} rows\n\
                              and Manually sampled input has {len({manual['sample_length']})}. Autosampled blood files\n\
                              should have more rows than manually sampled input files. Check .bld inputs.")
        

    def scale_time_rename_columns(self):
        """
        Scales time info if it's not in seconds and renames dataframe column to 'time' instead of given column name in 
        .bld file. Renames radioactivity column to BIDS compliant column name if it's in units Bq/cc or  Bq/mL.
        """
        # scale time info to seconds if it's minutes
        time_scalar = 1.0
        for name, dataframe in self.blood_series.items():
            time_column_header_name = [header for header in dataframe.columns if 'sec' in str.lower(header)]
            if not time_column_header_name:
                time_column_header_name = [header for header in dataframe.columns if 'min' in str.lower(header)]
                time_scalar = 60.0

            if time_column_header_name and len(time_column_header_name) == 1:
                print("Before Rename")
                print(dataframe)
                dataframe.rename(columns={time_column_header_name[0]: 'time'}, inplace=True)
            else:
                raise Exception("Unable to locate time column in blood file, make sure input files are formatted "
                                "to include a single time column in minutes or seconds.")

            # scale the time column to seconds
            dataframe['time'] = dataframe['time']*time_scalar
            self.blood_series[name] = dataframe

            # locate radioactivity column
            radioactivity_column_header_name = [header for header in dataframe.columns if 'bq' and 'cc' in str.lower(header)]
            parent_fraction_column_header_name = [header for header in dataframe.columns if 'parent' in str.lower(header)]
            if radioactivity_column_header_name and len(time_column_header_name) == 1:
                sub_ml_for_cc =re.sub('cc', 'mL', radioactivity_column_header_name[0])
                extracted_units = re.search(r'\[(.*?)\]', sub_ml_for_cc)
                if 'plasma' in str.lower(radioactivity_column_header_name[0]):
                    second_column_name = 'plasma_radioactivity'
                elif 'whole' in str.lower(radioactivity_column_header_name[0]) or 'blood' in str.lower(radioactivity_column_header_name[0]):
                    second_column_name = 'whole_blood_radioactivity'
                print(extracted_units.group(1))
                if extracted_units:
                    self.units = extracted_units.group(1)
                    dataframe.rename(columns={radioactivity_column_header_name[0]: second_column_name}, inplace=True)
                else:
                    raise Exception("Unable to determine radioactivity entries from .bld column name. Column name/units must be in Bq/cc or Bq/mL")
            elif parent_fraction_column_header_name and len(parent_fraction_column_header_name) == 1:
                    dataframe.rename(columns={parent_fraction_column_header_name[0]: 'metabolite_parent_fraction'}, inplace=True)

            print(dataframe)

    def ask_recording_type(self, recording: str):
        """
        Prompt user about data collection to determine how data was collected for each
        measure. e.g. auto-sampled, manually drawn, or a combination of the two
        """
        how = None
        while how != 'a' or how != 'm':
            how = input(f"How was the {recording} data sampled?:\nEnter A for automatically or M for manually\n")
            if str.lower(how) == 'm':
                self.data_collection[recording] = 'manual'
                break
            elif str.lower(how) == 'a':
                self.data_collection[recording] = 'automatic'
                break
            elif str.lower(how) == 'y':
                self.data_collection[recording] = 'manual'
                warnings.warn(f"Recieved {how} as input, assuming input recieved from cli w/ '-y' option on bash/zsh etc, defaulting to manual input")
                break
            else:
                print(f"You entered {how}; please enter either M or A to exit this prompt")

File: spreadsheet_conversion/blood/pmod/convert_pmod_to_blood.m
function convert_pmod_to_blood(varargin)

% routine the converts bld files (i.e. pmod pkin ready tab or comma separated values)
% this convert data on a subject per subject basis
% !! it supports only file with two columns, not the 'composite' type !!
% !!     column headers as per PMOD recommendation is also expected   !!
%
% Background
% PMOD (https://www.pmod.com/web/) PKIN software works with two or three files
% https://www.pmod.com/files/download/v31/doc/pkin/2235.htm
% WholebloodActivity: overall blood activity
%      = tracer in the red blood cells and tracer metabolites, mostly in the plasma.
% Plasma: overall plasma activity (tracer and metabolites)
% ParentFraction: free unchanged tracer from plasma (ie 'after' interacting
%                 with tissue) relative to total blood - this is the input to the brain
%
% :format: - convert_pmod_to_blood(filesin,'type','both',addjson,'off')
%          - convert_pmod_to_blood(filesin,'type','both','outputname','name','jsonkey', 'jsonvalue')
%
% :param FileListIn: if no input is provided, a GUI pops up
%       filesin a cell array of files e.g. ParentFraction,Wholeblood, Plasma pmod files
%       ParentFraction and Wholeblood files must be present, Plasma is optional
%       'type' should be 'manual' 'autosampler' or 'both' to indicate the way data were collected
%              (file time stamp should be enough to figure out which is which)
% :param option: 'outputname' for the base name of file(s) to export
%                'addjson' to turn off the creation of the side json file ('on' by default)
%                any key and value pairs to go in the blood json file
%
% :returns FileListOut: tsv files files for BIDS
%        if 'addjson' is 'on' (default), also creates an associated json file
%
% .. code-block::
%
%    Examples: 
%             file1 = fullfile(fileparts(which('convert_pmod_to_blood.m')),'parent_pmodexample.bld');
%             file2 = fullfile(fileparts(which('convert_pmod_to_blood.m')),'plasma_pmodexample.bld');
%             file3 = fullfile(fileparts(which('convert_pmod_to_blood.m')),'wholeblood_pmodexample.bld');
%             convert_pmod_to_blood(file1,file2,file3,'type','both','outputname','sub01-',...
%                 'MetaboliteMethod','HPLC','MetaboliteRecoveryCorrectionApplied','false',...
%                 'DispersionCorrected','false');
%
%             convert_pmod_to_blood('myfile1.bld','myfile2.bld','type','manual','outputname','sub01-',...
%                 'MetaboliteMethod','HPLC','MetaboliteRecoveryCorrectionApplied','false',...
%                 'DispersionCorrected','false');
%
% .. note::
%
%    Cyril Pernet - NRU
%    ----------------------------
%    Copyright Open NeuroPET team

%% PET BIDS parameters
jsoncheck = contains(varargin,'json','IgnoreCase',true);
if any(jsoncheck)
    addjson = varargin{find(jsoncheck)+1};
else
    addjson = 'on';
end

if strcmpi(addjson,'on')
    current    = which('convert_pmod_to_blood.m');
    root       = current(1:strfind(current,'converter')+length('converter'));
    jsontoload = fullfile(root,['metadata' filesep 'blood_metadata.json']);
    if exist(jsontoload,'file')
        bloodmetadata = jsondecode(fileread(jsontoload));
        mandatory     = bloodmetadata.mandatory;
        recommended   = bloodmetadata.recommended;
        clear bloodmetadata
    else
        error('looking for %s, but the file is missing',jsontoload)
    end
    
    % check library
    if ~exist('jsonwrite.m', 'file')
        error(['JSONio library jsonwrite.m file was not found but is needed,', ...
            ' jsonwrite.m is part of https://github.com/gllmflndn/JSONio but can also be found in the ONP matlab converter folder']);
    end
end

%% deal with input file
if nargin == 0
    [filenames, pathnames] = uigetfile({'*.bld','speadsheet';'*.txt','text'}, 'Pick the pmod files (N=2 or 3)', 'multiselect','on');
    if isequal(filenames, 0) || isequal(pathnames, 0)
        disp('Selection cancelled');
        return
    else
        for i = length(filenames):-1:1
            filein{i} = fullfile(pathnames, filenames{i});
            fprintf('files selected:\n%s\n', filein{i});
        end
    end
    
    % 'type' should be 'manual' 'autosampler' or 'both'
    type = questdlg('What type of blood sampling was performed?', ...
        'blood sampling', ...
        'manual', 'autosampler', 'both', 'manual');
    if isempty(type)
        warning('selection aborted - exiting'); return
    end
    
    outputname = inputdlg('please input the base name for the files to save');
    if isempty(outputname)
        warning('name aborted - exiting'); return
    else
        outputname = fullfile(pathnames,cell2mat(outputname));
    end
    
else
    for i = 1:3
        if ~strcmpi(varargin{i},'type')
            if ~exist(varargin{i}, 'file')
                error('%s not found', varargin{i});
            else
                filein{i} = varargin{i}; %#ok<AGROW>
            end
        end
    end
    
    for v=1:nargin
        if strcmpi(varargin{v},'outputname')
            outputname = varargin{v+1};
        elseif strcmpi(varargin{v},'type')
            type = varargin{v+1};
        end
    end
end

% detect what's inside the selected file for clean attribution
% assume most users would have used modern excel like, let matlab deal with
% xlm stuff if any using xlsx - default back on text otherwise (it has to
% fail through xlsx because textread always work, even if the contend is
% garbage)

for i=length(filein):-1:1
    if strcmpi(filein{i}(end-3:end),'.txt')
        movefile(filein{i},[filein{i}(1:end-3) 'tsv']);
        datain{i} = readtable([filein{i}(1:end-3) 'tsv'],'FileType','text','VariableNamingRule','preserve');
        movefile([filein{i}(1:end-3) 'tsv'],filein{i}); % back to normal
    else
        movefile(filein{i},[filein{i}(1:end-3) 'xlsx']); % fool matlab
        try
            datain{i} = readtable([filein{i}(1:end-3) 'xlsx'],'FileType','spreadsheet','VariableNamingRule','preserve');
            movefile([filein{i}(1:end-3) 'xlsx'],filein{i}); % back to normal
        catch % xlsxerr
            movefile([filein{i}(1:end-3) 'xlsx'],[filein{i}(1:end-3) 'tsv']);
            datain{i} = readtable([filein{i}(1:end-3) 'tsv'],'FileType','text','VariableNamingRule','preserve');
            movefile([filein{i}(1:end-3) 'tsv'],filein{i}); % back to normal
        end
    end
end

for i=length(filein):-1:1
    if any(contains(datain{i}.Properties.VariableNames,'value','IgnoreCase',true))
        [~,name] = fileparts(filein{i});
        warning('no variable name in %s, inferring from file name',name)
        if any(contains(name,'Fraction','IgnoreCase',true))
            ParentFraction = datain{i};
        elseif any(contains(name,'Whole','IgnoreCase',true)) && ...
                ~contains(name,'Plasma','IgnoreCase',true)
            Wholeblood     = datain{i};
        elseif any(contains(name,'Plasma','IgnoreCase',true))
            Plasma         = datain{i};
            if contains(name,'Whole','IgnoreCase',true) && contains(name,'Ratio','IgnoreCase',true) && ...
                any(Plasma.(Plasma.Properties.VariableNames{2}) < 1)
                warning('Plasma values seem to be ratio to whole blood data are being converted, check the tsv file')
                multiplyplasma = true;
            else
                multiplyplasma = false;
            end
        end
    else
        if any(contains(datain{i}.Properties.VariableNames,'parent','IgnoreCase',true))
            ParentFraction = datain{i};
        elseif any(contains(datain{i}.Properties.VariableNames,{'whole-blood','whole blood'},'IgnoreCase',true))
            Wholeblood     = datain{i};
        elseif any(contains(datain{i}.Properties.VariableNames,'plasma','IgnoreCase',true))
            Plasma         = datain{i};
        end
    end
end
clear filename datain

if ~exist('Wholeblood','var')
    error('WholebloodActivity file missing / not recognized')
end

% dirty check of parent variable
if any(ParentFraction.(ParentFraction.Properties.VariableNames{2}) > 1) && ...
        ~exist('Plasma','var')
    warning('Parent fraction seen as > 1 ?? assuming this is plasma')
    Plasma = ParentFraction;
    clear ParentFraction
end

% fix the time info to the right format
if any(contains(Wholeblood.Properties.VariableNames,{'minutes','min'},'IgnoreCase',true))
    varname = Wholeblood.Properties.VariableNames(find(contains(Wholeblood.Properties.VariableNames,{'minutes','min'},'IgnoreCase',true)));
    Wholeblood.(cell2mat(varname)) = seconds(minutes(Wholeblood.(cell2mat(varname)))); % transform to seconds
    WBtime = Wholeblood.(cell2mat(varname));
elseif any(contains(Wholeblood.Properties.VariableNames,{'seconds','sec'},'IgnoreCase',true))
    varname = Wholeblood.Properties.VariableNames(find(contains(Wholeblood.Properties.VariableNames,{'seconds','sec'},'IgnoreCase',true)));
    WBtime = Wholeblood.(cell2mat(varname));
end

if exist('ParentFraction','var')
    if any(contains(ParentFraction.Properties.VariableNames,{'minutes','min'},'IgnoreCase',true))
        varname = ParentFraction.Properties.VariableNames(find(contains(ParentFraction.Properties.VariableNames,{'minutes','min'},'IgnoreCase',true)));
        ParentFraction.(cell2mat(varname)) = seconds(minutes(ParentFraction.(cell2mat(varname))));
        PFtime = ParentFraction.(cell2mat(varname));
    elseif any(contains(ParentFraction.Properties.VariableNames,{'seconds','sec'},'IgnoreCase',true))
        varname = ParentFraction.Properties.VariableNames(find(contains(ParentFraction.Properties.VariableNames,{'seconds','sec'},'IgnoreCase',true)));
        PFtime = ParentFraction.(cell2mat(varname));
    end
end

if exist('Plasma','var')
    if any(contains(Plasma.Properties.VariableNames,{'minutes','min'},'IgnoreCase',true))
        varname = Plasma.Properties.VariableNames(find(contains(Plasma.Properties.VariableNames,{'minutes','min'},'IgnoreCase',true)));
        Plasma.(cell2mat(varname)) = seconds(minutes(Plasma.(cell2mat(varname))));
        Ptime = Plasma.(cell2mat(varname));
    elseif any(contains(Plasma.Properties.VariableNames,{'seconds','sec'},'IgnoreCase',true))
        varname = Plasma.Properties.VariableNames(find(contains(Plasma.Properties.VariableNames,{'seconds','sec'},'IgnoreCase',true)));
        Ptime = Plasma.(cell2mat(varname));
    end
end

% check time information across files
% if all manual or autosampler, it make sense to have the same time points
if ~strcmpi(type,'both')
    if exist('PFtime','var')
        if length(PFtime) ~= length(WBtime) || ~all(PFtime == WBtime)
            error('ParentActivity and Whole blood have different time values - this seems impossible')
        end
    end

    if exist('Ptime','var')
        if ~all(Ptime == WBtime)
            error('Whole blood and Plasma have different time values - this seems impossible')
        end
    end

    if strcmpi(type,'autosampler')
        if exist('Plasma','var') || exist('Parent','var')
            error('sorry, the function does not expect to see plasma or parent fraction from autosampling')
       end
    end

else % mixed case - assume the autosampling has more data points
    % possible that pmod exported file mixing time info
    % we simply clean up the data
    if length(WBtime) == length(Ptime) && exist('PFtime','var')
        warning('time data between whole blood and plasma are mixed, trying to fix it based on parent fraction - do check tsv files')
        duplicates = arrayfun(@(x) find(WBtime == x), PFtime);
        if ~isempty(duplicates)
            Ptime                     = Ptime(duplicates);  
            notduplicates             = 1:length(WBtime);
            notduplicates(duplicates) = [];
            WBtime                    = WBtime(notduplicates);
        end
    elseif length(WBtime) > length(Ptime) && exist('PFtime','var')
        warning('time data between whole blood and plasma are mixed, trying to fix it based on parent fraction - do check tsv files')
        if length(Ptime) ~= length(PFtime)
            if PFtime(1) == WBtime(1) && PFtime(1) == 0
                Ptime = [0;Ptime];                 
            else
                error('Parent fraction and Plasma times do not match, we cannot figure out which time points to extract in whole blood')
            end
        end
        
        duplicates = arrayfun(@(x) find(WBtime == x), PFtime, 'UniformOutput', false);
        if ~isempty(duplicates)
            emptyduplicates = find(cellfun(@(x) isempty(x),duplicates)); 
            if ~isempty(emptyduplicates)
                % strategy 1, is the missing index between two adjacent
                % integer, if so just used the value e.g. 183, [], 185
                for e=1:length(emptyduplicates)
                    position = emptyduplicates(e);
                    if position > 1 || position < length(duplicates)
                        if duplicates{position-1}+2 == duplicates{position+1}
                            duplicates{position} = duplicates{position-1}+1;
                        end
                    end
                end
            end

           emptyduplicates = find(cellfun(@(x) isempty(x),duplicates)); 
           if ~isempty(emptyduplicates)
               % strategy 2, try figuring out the closest time
                warning('whole blood times do not match perfectly parent fraction times - using closest match')
                missingtimes = PFtime(emptyduplicates);
                for m=1:length(missingtimes)
                    timevalue = min(WBtime((WBtime - missingtimes(m))>=0));
                    duplicates{emptyduplicates(m)} = find(WBtime == timevalue);
                end
            end
         
            if iscell(duplicates)
                duplicates = cell2mat(duplicates);
            end
            notduplicates             = 1:length(WBtime);
            notduplicates(duplicates) = [];
            WBtime                    = WBtime(notduplicates);
        end
    else
        duplicates = [];
    end
end

% make _recording-autosampler_blood.tsv
% make _recording-manual_blood.tsv
disp('exporting to tsv')
if ~exist('outputname','var')
    outputname = fullfile(fileparts(filein{1}),'pmodconverted');
end

whole_blood_radioactivity  = Wholeblood.(Wholeblood.Properties.VariableNames{2});

if exist('ParentFraction','var')
    metabolite_parent_fraction = ParentFraction.(ParentFraction.Properties.VariableNames{2});
end

if exist('ParentFraction','var')
    plasma_radioactivity = Plasma.(Plasma.Properties.VariableNames{2});
    % Ptime adjusted adding time 0, therefore adjust data as well
    if exist('metabolite_parent_fraction','var')
        if Ptime(1) == 0 && length(plasma_radioactivity) == length(metabolite_parent_fraction)-1
            plasma_radioactivity = [0;plasma_radioactivity];
        end
    end
    % Turns out we have a ratio rather the actual concentration so remultiply
    if multiplyplasma
        if ~isempty(duplicates)
            plasma_radioactivity = plasma_radioactivity.*whole_blood_radioactivity(duplicates);
        else
            plasma_radioactivity = plasma_radioactivity.*whole_blood_radioactivity;
        end
    end
end

if strcmpi(type,'both')
    if exist('PFtime','var')
        if ~isempty(duplicates)
            if length(plasma_radioactivity) == length(PFtime)
                t = table(PFtime,whole_blood_radioactivity(duplicates),plasma_radioactivity,...
                    metabolite_parent_fraction,'VariableNames',{'time','whole_blood_radioactivity',...
                    'plasma_radioactivity','metabolite_parent_fraction'});
            else
                t = table(PFtime,whole_blood_radioactivity(duplicates),plasma_radioactivity(duplicates),...
                    metabolite_parent_fraction,'VariableNames',{'time','whole_blood_radioactivity',...
                    'plasma_radioactivity','metabolite_parent_fraction'});
            end
        else
            t = table(PFtime,whole_blood_radioactivity,plasma_radioactivity,...
                metabolite_parent_fraction,'VariableNames',{'time','whole_blood_radioactivity',...
                'plasma_radioactivity','metabolite_parent_fraction'});
        end
    else
        t = table(WBtime,whole_blood_radioactivity,plasma_radioactivity,...
            'VariableNames',{'time','whole_blood_radioactivity','plasma_radioactivity'});
    end
    tsvname = [outputname '_recording-manual_blood.tsv'];
    writetable(t, tsvname, 'FileType', 'text', 'Delimiter', '\t');

    if exist('notduplicates','var')
        t = table(WBtime,whole_blood_radioactivity(notduplicates),...
            'VariableNames',{'time','whole_blood_radioactivity'});
    else
        t = table(WBtime,whole_blood_radioactivity,...
            'VariableNames',{'time','whole_blood_radioactivity'});
    end
    tsvname = [outputname '_recording-autosampler_blood.tsv'];
    writetable(t, tsvname, 'FileType', 'text', 'Delimiter', '\t');

else
    time  = Wholeblood.(Wholeblood.Properties.VariableNames{1});
    if exist('Plasma','var')
        plasma_radioactivity = Plasma.(Plasma.Properties.VariableNames{2});
        if exist('metabolite_parent_fraction','var')
            t = table(time,whole_blood_radioactivity,plasma_radioactivity,metabolite_parent_fraction,...
                'VariableNames',{'time','whole_blood_radioactivity','plasma_radioactivity','metabolite_parent_fraction'});
        else
            t = table(time,whole_blood_radioactivity,plasma_radioactivity,...
                'VariableNames',{'time','whole_blood_radioactivity','plasma_radioactivity'});
        end
    else
        t = table(time,whole_blood_radioactivity,metabolite_parent_fraction,...
            'VariableNames',{'time','whole_blood_radioactivity','metabolite_parent_fraction'});
    end
    tsvname = [outputname '_recording-' type '_blood.tsv'];
    writetable(t, tsvname, 'FileType', 'text', 'Delimiter', '\t');
end
 
%% export json

if strcmpi(addjson,'on')
    
    for v=1:nargin % since we have Parent those should be there
        if strcmpi(varargin{v},'MetaboliteMethod')
            MetaboliteMethod = varargin{v+1};
        elseif strcmpi(varargin{v},'MetaboliteRecoveryCorrectionApplied')
            MetaboliteRecoveryCorrectionApplied = varargin{v+1};
        elseif strcmpi(varargin{v},'DispersionCorrected')
            DispersionCorrected = varargin{v+1};
        end
    end
    
    info.WholeBloodAvail = "true"; % the function would have not run otherwise
    info.MetaboliteAvail = 'true';
    if exist('MetaboliteMethod','var')
        info.MetaboliteMethod = MetaboliteMethod;
    else
        warning('Parent fraction is available, but input method is not specified, which is not BIDS compliant')
    end
    
    if exist('MetaboliteRecoveryCorrectionApplied','var')
        info.MetaboliteRecoveryCorrectionApplied = MetaboliteRecoveryCorrectionApplied;
    else
        warning('Parent fraction is available, but there is no information if Recovery Correction was applied, which is not BIDS compliant')
    end
       
    if exist('DispersionCorrected','var')
        info.DispersionCorrected = DispersionCorrected;
    else
        warning('Parent fraction is available, but there is no information if DispersionCorrected was applied, which is not BIDS compliant')
    end
    
    info.time.Description                       = 'Time in relation to time zero defined by the _pet.json';
    info.time.Units                             = 's';
    info.whole_blood_radioactivity.Description  = 'Radioactivity in whole blood samples.';
    info.whole_blood_radioactivity.Units        = 'kBq/mL';
    info.metabolite_parent_fraction.Description = 'Parent fraction of the radiotracer';
    info.metabolite_parent_fraction.Units       = 'arbitrary';
    if exist('Plasma','var')
        info.PlasmaAvail                        = 'true';
        info.plasma_radioactivity.Description   = 'Radioactivity in plasma samples';
        info.plasma_radioactivity.Units         = 'kBq/mL';
    end
    
    for v=1:nargin 
        if any(strcmpi(varargin{v},recommended))
            index = find(strcmpi(varargin{v},recommended));
            info.(recommended{index}) = varargin{v+1}; %#ok<FNDSB>
        end
    end
    
    if ~strcmpi(type,'both')
       jsonname = [outputname '_recording-' type '_blood.json'];
    else
       jsonname = [outputname '_blood.json'];
    end
    jsonwrite(jsonname,info,'prettyprint','true');
   
    % run a last check
    for f=length(mandatory):-1:1
        M(f) = any(contains(mandatory{f}, fieldnames(info)));
    end
    if sum(M) ~= length(mandatory)
        index = find(M==0);
        for m=1:length(index)
            if iscell(mandatory{index(m)})
                tmp = mandatory{index(m)};
                for t=1:length(tmp)
                    warning('missing %s',tmp{t})
                end
            else
                warning('missing %s',mandatory{index(m)})
            end
        end
        warning('the json file created is NOT BIDS compliant, mandatory fields are missing')
    end
end


